rewrite-core/src/main/java/org/openrewrite/polyglot/Polyglot.java

```diff
@@ -0,0 +1,456 @@
+ *  <p>
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *  <p>
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.openrewrite.polyglot;
+
+import com.fasterxml.jackson.annotation.JsonIdentityInfo;
+import com.fasterxml.jackson.annotation.ObjectIdGenerators;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.With;
+import org.graalvm.polyglot.Value;
+import org.openrewrite.SourceFile;
+import org.openrewrite.Tree;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.marker.Marker;
+import org.openrewrite.marker.Markers;
+
+import java.io.Serializable;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static java.util.stream.Collectors.toList;
+import static org.openrewrite.Tree.randomId;
+
+
+@SuppressWarnings("unchecked")
+@JsonIdentityInfo(generator = ObjectIdGenerators.IntSequenceGenerator.class, property = "@ref")
+public interface Polyglot extends Serializable, Tree {
+
+    UUID getId();
+
+    Value getValue();
+
+    default <T> Optional<T> as(Class<T> targetType) {
+        try {
+            return Optional.ofNullable(getValue().as(targetType));
+        } catch (ClassCastException e) {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    default <R extends Tree, P> @Nullable R accept(TreeVisitor<R, P> v, P p) {
+        return v instanceof PolyglotVisitor ? (R) acceptPolyglot((PolyglotVisitor<P>) v, p) : v.defaultValue(this, p);
+    }
+
+    @Nullable
+    default <P> Polyglot acceptPolyglot(PolyglotVisitor<P> pv, P p) {
+        return pv.defaultValue(this, p);
+    }
+
+    @Override
+    default <P> boolean isAcceptable(TreeVisitor<?, P> v, P p) {
+        return v instanceof PolyglotVisitor;
+    }
+
+    interface Member extends Polyglot {
+        Markers getMarkers();
+
+        <P extends Polyglot> P withMarkers(Markers markers);
+
+        String getName();
+
+        Member withValue(Value value);
+
+        @Override
+        default @Nullable <P> Polyglot acceptPolyglot(PolyglotVisitor<P> pv, P p) {
+            return pv.visitMember(this, p);
+        }
+    }
+
+    @lombok.Value
+    @EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)
+    @With
+    class ValueMarker implements Marker {
+        @EqualsAndHashCode.Include
+        UUID id;
+
+        Value value;
+    }
+
+    @lombok.Value
+    @EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)
+    @With
+    class Source implements Polyglot, SourceFile {
```

Comment:
`Source` is a very similar name to `SourceFile`. Maybe this should be `PolyglotSource` or `PolyglotSourceFile`?

Reply:
Good point. Will probably use `PolyglotSource`.